File: homework-solutions/Group 14/beam.cpp
#include <bits/stdc++.h>
using namespace std;

struct PathState {
    vector<int> path;   // đường đi hiện tại
    long long visitedMask; // đánh dấu các đỉnh đã đi bằng bitmask
};

mt19937 rng(time(NULL));

int randint(int l, int r) {
    return uniform_int_distribution<int>(l, r) (rng);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;
    vector<vector<int>> g(N + 1);
    for (int i = 0; i < M; i++) {
        int u, v;
        cin >> u >> v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    if (N == 1) {
        cout << "0\n1\n";
        return 0;
    }

    // quick connectivity check
    vector<bool> reach(N + 1, false);
    queue<int> q;
    q.push(1); reach[1] = true;
    while (!q.empty()) {
        int u = q.front(); q.pop();
        for (int v : g[u]) if (!reach[v]) { reach[v] = true; q.push(v); }
    }
    if (!reach[N]) { cout << "-1\n"; return 0; }

    int beamWidth = 50000;
    vector<PathState> beam;
    PathState start; 
    start.path.push_back(1); start.visitedMask = 1LL << 0; // vertex 1
    beam.push_back(start);

    int bestLen = -1;
    vector<int> bestPath;

    while (!beam.empty()) {
        vector<PathState> nextBeam;

        for (auto &state : beam) {
            int u = state.path.back();
            if (u == N) {
                if ((int)state.path.size() > bestLen) {
                    bestLen = state.path.size();
                    bestPath = state.path;
                }
                continue;
            }
            for (int v : g[u]) {
                long long bit = 1LL << (v - 1);
                if (state.visitedMask & bit) continue; // đã đi rồi
                PathState ns;
                ns.path = state.path;
                ns.path.push_back(v);
                ns.visitedMask = state.visitedMask | bit;
                nextBeam.push_back(ns);
            }
        }

        sort(nextBeam.begin(), nextBeam.end(), [](const PathState &a, const PathState &b){
            if(a.path.size() != b.path.size()) return a.path.size() > b.path.size();
            if(a.visitedMask !=  b.visitedMask) return a.visitedMask < b.visitedMask;
            return a.path.back() < b.path.back();
        });
        for(int i = 1, j = 0; i < (int) nextBeam.size(); ++i) {
            if(nextBeam[j].visitedMask == nextBeam[i].visitedMask && nextBeam[j].path.back() == nextBeam[i].path.back()) {
                continue;
            } else nextBeam[++j] = nextBeam[i];
        }
        if ((int)nextBeam.size() > beamWidth) nextBeam.resize(beamWidth);

        beam = move(nextBeam);
    }

    start.path = {1};
    start.visitedMask = 1LL << 0; // vertex 1
    beamWidth = 10000;

    if (bestLen == -1) cout << "-1\n";
    else {
        cout << bestLen - 1 << "\n";
        for (int i = 0; i < bestPath.size(); i++) {
            if (i) cout << " ";
            cout << bestPath[i];
        }
        cout << "\n";
    }
    return 0;
}

File: homework-solutions/Group 14/beam.py
import sys
import threading
from collections import deque

def main():
    data = sys.stdin.read().strip().split()
    it = iter(data)
    N = int(next(it))
    M = int(next(it))

    g = [[] for _ in range(N + 1)]
    for _ in range(M):
        u = int(next(it))
        v = int(next(it))
        g[u].append(v)
        g[v].append(u)

    if N == 1:
        print("0")
        print("1")
        return

    reach = [False] * (N + 1)
    q = deque()
    q.append(1)
    reach[1] = True

    while q:
        u = q.popleft()
        for v in g[u]:
            if not reach[v]:
                reach[v] = True
                q.append(v)

    if not reach[N]:
        print("-1")
        return

    class PathState:
        __slots__ = ("path", "visitedMask")
        def __init__(self, path, visitedMask):
            self.path = path
            self.visitedMask = visitedMask

    beamWidth = 10000
    beam = [PathState([1], 1 << 0)]

    bestLen = -1
    bestPath = []

    while beam:
        nextBeam = []

        for state in beam:
            u = state.path[-1]
            if u == N:
                if len(state.path) > bestLen:
                    bestLen = len(state.path)
                    bestPath = state.path[:]
                continue

            for v in g[u]:
                bit = 1 << (v - 1)
                if state.visitedMask & bit:
                    continue
                new_path = state.path + [v]
                new_mask = state.visitedMask | bit
                nextBeam.append(PathState(new_path, new_mask))

        nextBeam.sort(key=lambda s: (-len(s.path), s.visitedMask, s.path[-1]))

        ptr = 0
        for i in range(len(nextBeam)):
            if i == 0:
                continue
            if nextBeam[ptr].visitedMask == nextBeam[i].visitedMask and nextBeam[ptr].path[-1] == nextBeam[i].path[-1]:
                continue
            ptr = ptr + 1
            nextBeam[ptr] = nextBeam[i]

        while len(nextBeam) > beamWidth:
            nextBeam.pop()

        beam = nextBeam

    if bestLen == -1:
        print("-1")
    else:
        print(bestLen - 1)
        print(" ".join(map(str, bestPath)))


if __name__ == "__main__":
    main()
